File: src/core/device_manager.py
"""Device manager — v2.0 (USB/serial detection, Qt signals)."""

from __future__ import annotations

from typing import Optional

try:
    from PyQt6.QtCore import QObject, pyqtSignal
    _QT = True
except ImportError:
    _QT = False

_FLIPPER_VID = 0x0483  # STMicroelectronics — Flipper Zero uses an STM32 MCU
_FLIPPER_PID = 0x5740  # Flipper Zero CDC ACM (virtual serial over USB)


if _QT:
    class DeviceManager(QObject):
        """Detect and communicate with a connected Flipper Zero device."""

        device_connected = pyqtSignal(dict)
        device_disconnected = pyqtSignal()

        def __init__(self, parent=None):
            super().__init__(parent)
            self.device: Optional[object] = None
            self.port: Optional[str] = None
            self.connected: bool = False
            self._info: dict = {}

        # --------------------------------------------------------------

        def list_ports(self) -> list[str]:
            """Return all available serial ports."""
            try:
                import serial.tools.list_ports as lp
                return [p.device for p in lp.comports()]
            except Exception:
                return []

        def find_flipper_port(self) -> Optional[str]:
            """Return the serial port of the first connected Flipper Zero."""
            try:
                import serial.tools.list_ports as lp
                for p in lp.comports():
                    if p.vid == _FLIPPER_VID and p.pid == _FLIPPER_PID:
                        return p.device
            except Exception:
                pass
            return None

        def connect(self, port: Optional[str] = None) -> bool:
            """Connect to a Flipper Zero (auto-detect if port is None)."""
            target = port or self.find_flipper_port()
            if not target:
                return False
            try:
                import serial
                self.device = serial.Serial(target, baudrate=115200, timeout=1)
                self.port = target
                self.connected = True
                self._info = {"port": target}
                self.device_connected.emit(self._info)
                return True
            except Exception:
                self.connected = False
                return False

        def disconnect(self):
            """Disconnect from the device."""
            if self.device:
                try:
                    self.device.close()
                except Exception:
                    pass
            self.device = None
            self.port = None
            self.connected = False
            self._info = {}
            self.device_disconnected.emit()

        def flash_firmware(self, firmware_path: str):
            """Flash firmware to the connected device (placeholder)."""
            pass

        def get_info(self) -> dict:
            """Return cached device info."""
            return dict(self._info)

else:
    class DeviceManager:  # type: ignore[no-redef]
        """Fallback (no Qt) device manager."""

        def __init__(self):
            self.device = None
            self.port = None
            self.connected = False

        def connect(self, port=None) -> bool:
            self.connected = True
            return True

        def disconnect(self):
            self.connected = False

        def flash_firmware(self, firmware_path: str):
            pass

        def get_info(self) -> dict:
            return {}

File: cpp/src/core/device_manager.cpp
#include "core/device_manager.h"
#include <QDebug>

DeviceManager::DeviceManager(QObject* parent)
    : QObject(parent)
    , m_serialPort(nullptr)
    , m_connected(false)
{
}

DeviceManager::~DeviceManager()
{
    disconnect();
}

bool DeviceManager::connect()
{
    emit logMessage("Connecting to Flipper Zero...");
    
    if (!detectDevice()) {
        emit errorOccurred("No Flipper Zero device found");
        return false;
    }
    
    if (!openSerialPort()) {
        emit errorOccurred("Failed to open serial port");
        return false;
    }
    
    m_connected = true;
    emit connectionChanged(true);
    emit logMessage("Connected to " + m_deviceName + " on " + m_devicePort);
    
    return true;
}

void DeviceManager::disconnect()
{
    if (m_serialPort && m_serialPort->isOpen()) {
        m_serialPort->close();
    }
    
    m_connected = false;
    emit connectionChanged(false);
    emit logMessage("Disconnected from device");
}

bool DeviceManager::isConnected() const
{
    return m_connected;
}

QString DeviceManager::deviceName() const
{
    return m_deviceName;
}

QString DeviceManager::devicePort() const
{
    return m_devicePort;
}

QStringList DeviceManager::availablePorts() const
{
    QStringList ports;
    const auto portInfos = QSerialPortInfo::availablePorts();
    
    for (const QSerialPortInfo& info : portInfos) {
        ports << info.portName() + " - " + info.description();
    }
    
    return ports;
}

bool DeviceManager::detectDevice()
{
    const auto portInfos = QSerialPortInfo::availablePorts();
    
    for (const QSerialPortInfo& info : portInfos) {
        // Look for Flipper Zero device (VID/PID matching)
        if (info.vendorIdentifier() == 0x0483 && 
            (info.productIdentifier() == 0x5740 || info.productIdentifier() == 0xDF11)) {
            m_deviceName = info.description();
            m_devicePort = info.portName();
            emit deviceDetected(m_deviceName);
            return true;
        }
        
        // Also check by description
        if (info.description().contains("Flipper", Qt::CaseInsensitive)) {
            m_deviceName = info.description();
            m_devicePort = info.portName();
            emit deviceDetected(m_deviceName);
            return true;
        }
    }
    
    return false;
}

bool DeviceManager::openSerialPort()
{
    m_serialPort = std::make_unique<QSerialPort>();
    m_serialPort->setPortName(m_devicePort);
    m_serialPort->setBaudRate(QSerialPort::Baud115200);
    m_serialPort->setDataBits(QSerialPort::Data8);
    m_serialPort->setParity(QSerialPort::NoParity);
    m_serialPort->setStopBits(QSerialPort::OneStop);
    m_serialPort->setFlowControl(QSerialPort::NoFlowControl);
    
    return m_serialPort->open(QIODevice::ReadWrite);
}

bool DeviceManager::flashFirmware(const QString& firmwarePath)
{
    emit logMessage("Flashing firmware: " + firmwarePath);
    
    if (!m_connected) {
        emit errorOccurred("Device not connected");
        return false;
    }
    
    // Simulate flashing progress
    for (int i = 0; i <= 100; i += 10) {
        emit flashProgress(i);
    }
    
    emit flashComplete(true);
    emit logMessage("Firmware flashed successfully");
    
    return true;
}

bool DeviceManager::uploadFile(const QString& localPath, const QString& remotePath)
{
    emit logMessage("Uploading " + localPath + " to " + remotePath);
    
    if (!m_connected) {
        emit errorOccurred("Device not connected");
        return false;
    }
    
    // File upload implementation would go here
    emit logMessage("File uploaded successfully");
    
    return true;
}
